bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/DateExpression.java

```diff
@@ -0,0 +1,171 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.common;
+
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.TimeZone;
+
+import javax.xml.bind.DatatypeConverter;
+
+import org.eclipse.smarthome.core.common.DateExpression.DateExpressionPart;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <code>DateExpression</code> is an implementation of {@link Expression} that provides a parser and evaluator for for
+ * ISO8601 date expressions (https://en.wikipedia.org/wiki/ISO_8601). ISO8601 expressions provide the ability to specify
+ * simple yet precise dates.
+ * <P>
+ * ISO8601 expressions are comprised of 7 required fields described as follows:
+ *
+ * YYYY-MM-DDThh:mm:ss.sTZD
+ *
+ * where
+ *
+ * YYYY = four-digit year
+ * MM = two-digit month (01=January, etc.)
+ * DD = two-digit day of month (01 through 31)
+ * hh = two digits of hour (00 through 23) (am/pm NOT allowed)
+ * mm = two digits of minute (00 through 59)
+ * ss = two digits of second (00 through 59)
+ * s = one or more digits representing a decimal fraction of a second
+ * TZD = time zone designator (Z or +hh:mm or -hh:mm)
+ *
+ * @author Karel Goderis - Intial Contribution
+ *
+ */
+public class DateExpression extends AbstractExpression<DateExpressionPart> {
+
+    private final static Logger logger = LoggerFactory.getLogger(DateExpression.class);
+
+    public DateExpression(final String date) throws ParseException {
+        this(date, Calendar.getInstance().getTime(), TimeZone.getDefault());
+    }
+
+    public DateExpression(final String dateExpression, final Date startTime, final TimeZone zone)
+            throws ParseException {
+        super(dateExpression, "", startTime, zone, 0);
+    }
+
+    @Override
+    public boolean isSatisfiedBy(Date date) {
+        if (date == null) {
+            throw new IllegalArgumentException("Date cannot be null");
+        }
+        Calendar testDateCal = Calendar.getInstance(getTimeZone());
+        testDateCal.setTime(date);
+        testDateCal.set(Calendar.MILLISECOND, 0);
+        Date originalDate = testDateCal.getTime();
+
+        testDateCal.add(Calendar.SECOND, -1);
+
+        Date timeAfter = getTimeAfter(testDateCal.getTime());
+
+        return ((timeAfter != null) && (timeAfter.equals(originalDate)));
+    }
+
+    public static boolean isValidExpression(String dateExpression) {
+        try {
+            new DateExpression(dateExpression);
+        } catch (ParseException pe) {
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public Date getFinalFireTime() {
+        if (candidates.isEmpty()) {
```

Comment:
This block is repeated a couple of times - I think it would be good to factor it out into the abstract base class